File: smallest_divisor.rb
# find smallest divisor of an integer
# adapted from SICP section 1.2

def square(n)
  n * n
end  

def smallest_divisor(n)
  find_divisor(n,2)
end

def find_divisor(n,test_divisor)
  case
  when square(test_divisor) > n
    n
  when divides(test_divisor,n)
    test_divisor
  else
    find_divisor(n,(test_divisor+1))
  end
end

def divides(a,b)
  b % a == 0
end

def prime(n)
  smallest_divisor(n) == n
end

(1..50).each do |number|
  puts "#{number} is prime: #{prime(number)}"
end  

File: smallest_divisor.php
<?php

function square($n) {
  return ($n * $n);
}

function smallest_divisor($n) {
  return find_divisor($n,2);
}

function divides($a,$b) {
  if ($b % $a == 0) {
    return 1;
  } else {
    return 0;
  }
}

function find_divisor($n,$test_divisor) {
  if (square($test_divisor) > $n) {
    return $n;
  } else if (divides($test_divisor,$n) == 1) {
    return $test_divisor;
  } else {
    return find_divisor($n,($test_divisor+1));
  }
}

function is_prime($n) {
  if (smallest_divisor($n) == $n) {
    return 1;
  } else {
    return 0;
  }
}

for ($i = 1; $i <= 50; $i++) {
  if (is_prime($i) == 1) {
    echo "$i is prime\n";
  }
}


